### PR TITLE
[V4] Handle graphql subscriptions

### PR DESCRIPTION
**Pull request by Convly**
### What does it do

Allow to declare and use subscription with the GraphQL plugin.
The policies, middleware & authorization configurations are also supported.

### Example

Adds a `onPublishCategory` subscription that sends a category every time one is published

`/src/index.js`

```js
'use strict';

const { on } = require('events');
const { prop, propEq, identity } = require('lodash/fp');

module.exports = {
  /**
   * An asynchronous register function that runs before
   * your application is initialized.
   *
   * This gives you an opportunity to extend code.
   */
  register({ strapi }) {
    // Generator which yield when a category is published
    async function* onPublishCategory() {
      for await (const items of on(strapi.eventHub, 'entry.publish')) {
        const categories = items.filter(propEq('model', 'category')).map(prop('entry'));

        for (const category of categories) {
          yield category;
        }
      }
    }

    strapi
      .plugin('graphql')
      .service('extension')
      .use(({ nexus }) => ({
        // Register the subscription
        types: [
          nexus.subscriptionType({
            definition(t) {
              t.boolean('onPublishCategory', {
                type: 'Category',
                subscribe: onPublishCategory,
                // Ideally, we could sanitize the result in the resolve
                // method instead of returning the entry unchanged
                resolve: identity,
              });
            },
          }),
        ],

        // Disable auth for the subscription
        resolversConfig: { 'Subscription.onPublishCategory': { auth: false } },
      }));
  },

  /**
   * An asynchronous bootstrap function that runs before
   * your application gets started.
   *
   * This gives you an opportunity to set up your data model,
   * run jobs, or perform some special logic.
   */
  bootstrap({ strapi }) {},

  /**
   * An asynchronous destroy function that runs before
   * your application gets shut down.
   *
   * This gives you an opportunity to gracefully stop services you run.
   */
  destroy({ strapi }) {},
};

```

### CODE REVIEW

**Review by Elhebert — APPROVED** [review]
LGTM 👌

File: packages/plugins/graphql/server/bootstrap.js [hunk 1]
```diff
@@ -61,12 +65,27 @@ module.exports = async ({ strapi }) => {
 
   const serverConfig = merge(defaultServerConfig, config('apolloServer', {}));
 
+  // Handle subscriptions
+  if (config('subscriptions', true)) {
```
Comment (alexandrebodin) [inline packages/plugins/graphql/server/bootstrap.js:69]: As we don't provide any feature we better disable it by default

File: packages/plugins/graphql/server/services/content-api/wrap-resolvers.js [hunk 2]
```diff
@@ -31,87 +31,94 @@ const wrapResolvers = ({ schema, strapi, extension = {} }) => {
...
-        const hasConfig = !isNil(authConfig);
-
-        const isAuthDisabled = authConfig === false;
-
-        if ((isMutationOrQuery || hasConfig) && !isAuthDisabled) {
-          try {
-            await strapi.auth.verify(authContext, authConfig);
-          } catch (error) {
-            throw new ForbiddenError();
+  const typesMaps = [schema.getTypeMap()];
```
Comment (alexandrebodin) [inline packages/plugins/graphql/server/services/content-api/wrap-resolvers.js:34]: as the subscription is a type are you sure it's not in this already ?

File: packages/plugins/graphql/server/services/content-api/wrap-resolvers.js [hunk 3]
```diff
@@ -31,87 +31,94 @@ const wrapResolvers = ({ schema, strapi, extension = {} }) => {
...
+
+  const subscriptionType = schema.getSubscriptionType();
+  if (subscriptionType) {
+    typesMaps.push(subscriptionType.getFields());
```
Comment (alexandrebodin) [inline packages/plugins/graphql/server/services/content-api/wrap-resolvers.js:38]: Shouldn't the entire subscription type be passed in the typesMaps? Should the function handle the fields correctly already?

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/plugins/graphql/server/services/content-api/wrap-resolvers.js (hunk 2 above)
Comment (Convly) [reply packages/plugins/graphql/server/services/content-api/wrap-resolvers.js:34]: Good catch

File: packages/plugins/graphql/server/services/content-api/wrap-resolvers.js (hunk 3 above)
Comment (Convly) [reply packages/plugins/graphql/server/services/content-api/wrap-resolvers.js:38]: Don't need it as the comment above makes it irrelevant

File: packages/plugins/graphql/server/services/builders/utils.js [hunk 4]
```diff
@@ -95,16 +95,15 @@ module.exports = ({ strapi }) => {
...
       const { pagination = {}, filters = {} } = args;
 
       // Init
       const newArgs = omit(['pagination', 'filters'], args);
 
       // Pagination
       if (usePagination) {
-        const defaultLimit = strapi.plugin('graphql').config('defaultLimit');
-        const maxLimit = strapi.plugin('graphql').config('maxLimit', -1);
+        const { defaultLimit, maxLimit } = config;
```
Comment (alexandrebodin) [inline packages/plugins/graphql/server/services/builders/utils.js:106]: no more defaults ?

File: packages/plugins/graphql/server/services/utils/config.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+/**
+ * GraphQL config helper with consistent defaults values
+ */
+module.exports = ({ strapi }) => {
```
Comment (alexandrebodin) [inline packages/plugins/graphql/server/services/utils/config.js:6]: why not put that in the config default values instead ?

Comment (alexandrebodin) [reply packages/plugins/graphql/server/services/utils/config.js:6]: It doesn't look like it's necessary

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/plugins/graphql/server/services/builders/utils.js (hunk 4 above)
Comment (Convly) [reply packages/plugins/graphql/server/services/builders/utils.js:106]: Handled in the new graphql config service

File: packages/plugins/graphql/server/services/utils/config.js (hunk 5 above)
Comment (Convly) [reply packages/plugins/graphql/server/services/utils/config.js]: didn't know it existed, will do it

**Review by alexandrebodin — APPROVED** [review]
LGTM

**Comment by groger** [comment]
hello, 

I was wondering if anyone tested this? I keep getting errors when testing this on the Strapi playground. But queries and mutations are working perfectly fine.
"error": "Could not connect to websocket endpoint ws://localhost:1337/graphql. Please check if the endpoint url is correct."

Thank you

**Comment by Convly** [comment]
Hello @groger, did you add `subscriptions: true` to your `graphql` configuration in the `plugins.js` config file of your project?

**Comment by harshmalvi** [comment]
@groger I'm getting the same issues.
"error": "Could not connect to websocket endpoint ws://localhost:1337/graphql. Please check if the endpoint url is correct."

is that fixed?

**Pull request merged**